### PR TITLE
core: refactor random region

### PR DESCRIPTION
**Pull request by disksing**
### What problem does this PR solve? 
In the `placement rules` implementation, the `RegionOption` which passed to `RandXxxRegion` needs to call `cluster.GetStores()`. It meets deadlock problem because it first needs lock `BasicCluster` before get into `regions.RandXxxRegion`.

### What is changed and how it works?
Check `RegionOption` in `BasicCluster`. The procedure is:
1. Lock `BasicCluster`
2. Get 10 random regions without check `RegionOption`.
3. Release Lock.
4. Check `RegionOption` (during the check it may require Lock again)

### Check List 

Tests 
 - Unit test

### CODE REVIEW

File: server/core/region_test.go [hunk 1]
```diff
@@ -134,12 +134,18 @@ func (*testRegionKey) TestRegionKey(c *C) {
 func BenchmarkRandomRegion(b *testing.B) {
 	regions := NewRegionsInfo()
 	for i := 0; i < 5000000; i++ {
-		item := &RegionInfo{meta: &metapb.Region{StartKey: []byte(fmt.Sprintf("%20d", i)), EndKey: []byte(fmt.Sprintf("%20d", i+1))}}
-		regions.AddRegion(item)
+		peer := &metapb.Peer{StoreId: 1, Id: uint64(i + 1)}
```
Comment (nolouch) [inline server/core/region_test.go]: The benchmark may drop 10x of `RandXXXRegion ` :(

File: server/core/basic_cluster.go [hunk 2]
```diff
@@ -151,25 +152,47 @@ func (bc *BasicCluster) UpdateStoreStatus(storeID uint64, leaderCount int, regio
...
 }
 
+const randomRegionMaxRetry = 10
+
 // RandFollowerRegion returns a random region that has a follower on the store.
 func (bc *BasicCluster) RandFollowerRegion(storeID uint64, opts ...RegionOption) *RegionInfo {
 	bc.RLock()
-	defer bc.RUnlock()
-	return bc.Regions.RandFollowerRegion(storeID, opts...)
+	regions := bc.Regions.RandFollowerRegions(storeID, randomRegionMaxRetry)
```
Comment (lhy1024) [inline server/core/basic_cluster.go]: BTW, we need to ensure there is one or more region passed check option

Comment (disksing) [reply server/core/basic_cluster.go]: Can't ensure that. It's possible that no region can pass checkers. Old implementation only try 10 times too.

**Pull request merged**